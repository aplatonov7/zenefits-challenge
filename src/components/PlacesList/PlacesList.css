.PlacesList__list {
  background-color: white;
  overflow: auto;
  flex-grow: 1;
  box-shadow: 0 1px 4px rgba(0, 0, 0, 0.3);
}

.PlacesList__item {
  border-bottom: 1px solid #ddd;
  padding: 1.5rem 1.75rem;
  display: flex;
  width: 100%;
  text-align: left;
  cursor: pointer;    
}

.PlacesList__item:hover {
  background-color: #eee; 
}

.PlacesList__item-info {
  flex: 1;
}

.PlacesList__item-name {
  font-size: 1.15rem;
  line-height: 2;
}

.PlacesList__item-address {
  color: #666;
}

.PlacesList__item-preview {
  height: 100px;
  background-size: cover;
  background-position: center;
  flex: 0 0 100px;
  margin-left: 1rem;
  background-color: #ccc;
  display: flex;
  align-items: center;
  justify-content: center;
  color: #333;
  font-size: 1.2rem;
  text-align: center;
  padding: 0.5rem;
}

.PlacesList__item-open {
  line-height: 2;
  color: #c00;
}

.PlacesList__no-results {
  text-align: center;
  background-color: white;
  flex-grow: 1;
  font-size: 1.5rem;
  padding: 1.5rem 0;
  box-shadow: 0 1px 4px rgba(0, 0, 0, 0.3);
}