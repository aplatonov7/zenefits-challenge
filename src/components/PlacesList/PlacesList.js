import React from 'react'
import PropTypes from 'prop-types'

import './PlacesList.css'

export const PlacesListItem = (props) => {
  const { name, preview, isOpen, address, onClick } = props
  return (
    <li>
      <button className="PlacesList__item" onClick={() => onClick(props)}>
        <div className="PlacesList__item-info">
          <div className="PlacesList__item-name">{name}</div>
          <div className="PlacesList__item-address">{address}</div>
          {isOpen && <div className="PlacesList__item-open">Open now</div>}
        </div>
        <div 
          className="PlacesList__item-preview" 
          style={{ backgroundImage: preview ? `url(${preview})` : 'none' }}
        >
          {preview == null && 'No Preview'}
        </div>
      </button>
    </li>
  )
}

PlacesListItem.propTypes = {
  address: PropTypes.string,
  isOpen: PropTypes.bool,
  name: PropTypes.string,
  preview: PropTypes.string,
  onClick: PropTypes.func,
}

const PlacesList = ({ items, onPlaceClick }) => items.length > 0 ? (
  <ul className="PlacesList__list">
    {items.map(item => <PlacesListItem onClick={onPlaceClick} key={item.id} {...item} />)}
  </ul>
) : (
  <div className="PlacesList__no-results">No results found</div>
)

PlacesList.propTypes = {
  items: PropTypes.array,
  onPlaceClick: PropTypes.func,
}

PlacesList.defaultProps = {
  items: [],
}

export default PlacesList