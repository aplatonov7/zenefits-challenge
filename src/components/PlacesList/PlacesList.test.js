import React from 'react'

import PlacesList, { PlacesListItem } from './PlacesList'

describe('PlacesList component', () => {
  let component

  beforeEach(() => {
    component = shallow(<PlacesList />)
  })
  
  it('Should match snapshot', () => {
    const props = {
      items: [{ id: '1' }, { id: '2' }, { id: '3' }],
      onPlaceCLick: () => {},
    }
    expect(renderer.create(<PlacesList {...props} />).toJSON()).toMatchSnapshot()
  })

  it('Should render PlacesListItem component for every element of items prop', () => {
    component.setProps({ items: [{ id: '1' }, { id: '2' }, { id: '3' }] })
    expect(component.find(PlacesListItem)).toHaveLength(3)
  })

  it('Should display "no result" response when items prop is empty', () => {
    component.setProps({ items: [] })
    expect(component.find('.PlacesList__no-results')).toHaveLength(1)
  })
})

describe('PlacesListItem', () => {
  let component

  beforeEach(() => {
    component = shallow(<PlacesListItem />)
  })

  it('Should match snapshot', () => {
    const props = {
      address: 'Testing address',
      isOpen: true,
      name: 'Some name',
      preview: 'some_url',
      onClick: () => {},
    }
    expect(renderer.create(<PlacesListItem {...props} />).toJSON()).toMatchSnapshot()
  })

  it('Should render address', () => {
    component.setProps({ address: 'Testing address' })
    expect(component.find('.PlacesList__item-address').text()).toBe('Testing address')
  })

  it('Should render name', () => {
    component.setProps({ name: 'Testing name' })
    expect(component.find('.PlacesList__item-name').text()).toBe('Testing name')
  })

  it('Should render preview image based on provided preview prop', () => {
    expect(component.find('.PlacesList__item-preview').text()).toBe('No Preview')
    component.setProps({ preview: 'some_url' })
    expect(component.find('.PlacesList__item-preview').text()).toBe('')
    expect(component.find('.PlacesList__item-preview').html().match(/style="([^"]*)"/i)[1])
      .toBe('background-image:url(some_url)')
  })

  it('Should render "Open now" label based on the isOpen prop', () => {
    component.setProps({ isOpen: false })
    expect(component.find('.PlacesList__item-open')).toHaveLength(0)
    component.setProps({ isOpen: true })
    expect(component.find('.PlacesList__item-open')).toHaveLength(1)
  })

  it('Should fire onClick prop on button click', () => {
    const mockFn = jest.fn()
    component.setProps({ onClick: mockFn })
    component.find('button').simulate('click')
    expect(mockFn).toBeCalled()
  })
})