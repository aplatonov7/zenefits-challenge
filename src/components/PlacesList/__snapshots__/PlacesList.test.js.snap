// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlacesList component Should match snapshot 1`] = `
<ul
  className="PlacesList__list"
>
  <li>
    <button
      className="PlacesList__item"
      onClick={[Function]}
    >
      <div
        className="PlacesList__item-info"
      >
        <div
          className="PlacesList__item-name"
        />
        <div
          className="PlacesList__item-address"
        />
      </div>
      <div
        className="PlacesList__item-preview"
        style={
          Object {
            "backgroundImage": "none",
          }
        }
      >
        No Preview
      </div>
    </button>
  </li>
  <li>
    <button
      className="PlacesList__item"
      onClick={[Function]}
    >
      <div
        className="PlacesList__item-info"
      >
        <div
          className="PlacesList__item-name"
        />
        <div
          className="PlacesList__item-address"
        />
      </div>
      <div
        className="PlacesList__item-preview"
        style={
          Object {
            "backgroundImage": "none",
          }
        }
      >
        No Preview
      </div>
    </button>
  </li>
  <li>
    <button
      className="PlacesList__item"
      onClick={[Function]}
    >
      <div
        className="PlacesList__item-info"
      >
        <div
          className="PlacesList__item-name"
        />
        <div
          className="PlacesList__item-address"
        />
      </div>
      <div
        className="PlacesList__item-preview"
        style={
          Object {
            "backgroundImage": "none",
          }
        }
      >
        No Preview
      </div>
    </button>
  </li>
</ul>
`;

exports[`PlacesListItem Should match snapshot 1`] = `
<li>
  <button
    className="PlacesList__item"
    onClick={[Function]}
  >
    <div
      className="PlacesList__item-info"
    >
      <div
        className="PlacesList__item-name"
      >
        Some name
      </div>
      <div
        className="PlacesList__item-address"
      >
        Testing address
      </div>
      <div
        className="PlacesList__item-open"
      >
        Open now
      </div>
    </div>
    <div
      className="PlacesList__item-preview"
      style={
        Object {
          "backgroundImage": "url(some_url)",
        }
      }
    />
  </button>
</li>
`;
